transcom/mymove
# B 18490 update mto v2 final destination sit int

## [B-18490](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A889876)

## Summary

This is a service object level update to restrict SIT final destination address from being updated by the `updateMTOShipment` endpoint. updateMTOShipment V1 is marked for deprecation as the V2 endpoint will remove the functionality to update a service item sit final destination address.

This is difficult to test as it is an edge case. [E-04539](https://www13.v1host.com/USTRANSCOM38/Epic.mvc/Summary?oidToken=Epic%3a834680) will eventually add in functionality that will be restricted by this endpoint, making this no longer an edge case. 

## Method

Rather than recreate the handlers and its dependent functions, I created a new V2 handler and utilized the same exact service objects. However, I added a new field to the mto shipment updater service object, taking in an "api" parameter. This allows the conditional logic of how to handle requests from different APIs, or in this case, it puts a 1 line function call behind an if condition of `if api == "prime"`. The function call put behind this if condition is `UpdateDestinationSITServiceItemsAddress`.

## Endpoint deprecation

Prime V1 `updateMTOShipment` is slated for deprecation on August 5th, 2024

## Test Setup

Please see how to import the latest postman collection [here](https://transcom.github.io/mymove-docs/docs/tools/postman/setting-up-postman#importing-a-collection).

See how to setup postman [here](https://transcom.github.io/mymove-docs/docs/tools/postman/setting-up-postman).

## How to test

Login as prime
Select any move
Create new PPM shipment, see image for specific details.

<img width="651" alt="Pasted Graphic 5" src="https://github.com/transcom/mymove/assets/136814362/d099824a-4f9e-49b9-bab8-6b54fa5106ba">

<img width="689" alt="Pasted Graphic 6" src="https://github.com/transcom/mymove/assets/136814362/40630dfc-0bb3-4c5f-ab3b-622897bb4cbe">

Scroll down to your newly created PPM shipment.

<img width="1051" alt="Pasted Graphic 7" src="https://github.com/transcom/mymove/assets/136814362/d87ce4ab-21f7-4b4c-bc18-aa3e36c60465">


Click Add Service Item
Populate for Destination SIT

<img width="598" alt="Pasted Graphic 9" src="https://github.com/transcom/mymove/assets/136814362/9d82ca67-0361-4a48-a9ce-f809697a6213">


Create the service item

Scroll down and grab your PPM shipment ID and eTAG

NOTE
As of this PR, it will create with a NULL final destination address id. See screenshot for proof

<img width="483" alt="Pasted Graphic 10" src="https://github.com/transcom/mymove/assets/136814362/534304c7-bf84-490f-ad51-a6b09fba0df1">

Go to postman

Navigate to prime api -> mto-shipments -> mtoShipmentID -> updateMTOShipment
Update URL to use the PPM shipment ID
Add If-Match header and input the eTag

<img width="619" alt="Pasted Graphic 8" src="https://github.com/transcom/mymove/assets/136814362/d81298a4-6ed6-4dca-a4a9-1f0412b0f971">


Use a body like this: 
```

{
  "scheduledPickupDate": "2024-02-05",
  "actualPickupDate": "2024-02-05",
  "firstAvailableDeliveryDate": "2024-02-05",
  "scheduledDeliveryDate": "2024-02-05",
  "actualDeliveryDate": "2024-02-05",
  "primeEstimatedWeight": 4500,
  "primeActualWeight": 4500,
  "destinationAddress": {
    "id": "972824c8-c083-4af8-8b76-450288abd2dd",
    "streetAddress1": "123 Cobble Ave",
    "streetAddress2": "Apartment 9000",
    "streetAddress3": "Montmârtre",
    "city": "Anytown",
    "state": "AL",
    "postalCode": "90210",
    "country": "USA"
  },
  "destinationType": "OTHER_THAN_AUTHORIZED",
  "shipmentType": "PPM",
  "diversion": false,
  "ppmShipment": {
    "expectedDepartureDate": "2024-02-05",
    "pickupPostalCode": "90210",
    "secondaryPickupPostalCode": "90210",
    "destinationPostalCode": "90210",
    "secondaryDestinationPostalCode": "90210",
    "sitExpected": true,
    "sitLocation": "ORIGIN",
    "sitEstimatedWeight": 2000,
    "sitEstimatedEntryDate": "2024-02-05",
    "sitEstimatedDepartureDate": "2024-02-05",
    "estimatedWeight": 4200,
    "hasProGear": true,
    "proGearWeight": 0,
    "spouseProGearWeight": 0
  }
}
```

If you submit this body and header to the V1 endpoint, then you can refresh your mto_service_items view and see that a sit dest final address id is now present within the database. If you submit the body and header to the V2 endpoint then the sit dest final address id will remain null.

V1 endpoint example: `https://primelocal:9443/prime/v1/mto-shipments/3d9f85fd-7f91-4b45-a434-9fdcc9d64e0a`
V2 endpoint example: `https://primelocal:9443/prime/v2/mto-shipments/3d9f85fd-7f91-4b45-a434-9fdcc9d64e0a`

Make sure if using postman environment variables that `{{baseURL}}` points to the correct version.

<img width="568" alt="Pasted Graphic 12" src="https://github.com/transcom/mymove/assets/136814362/07f37125-6754-42f4-ba44-64b8fd3d4aa9">

If you utilize the V2 endpoint of `updateMTOShipment` and sit dest final address id remains null for the service items, you have successfully tested this PR.
